src/KingSystem/Utils/Byaml/Byaml.cpp

```diff
@@ -0,0 +1,399 @@
+#include "KingSystem/Utils/Byaml/ByamlHashIter.h"
+#include "KingSystem/Utils/Byaml/ByamlLocal.h"
+#include "KingSystem/Utils/Byaml/ByamlStringTableIter.h"
+#include "math/seadMathCalcCommon.h"
+
+namespace al {
+
+ByamlIter::ByamlIter() : mData(nullptr), mRootNode(nullptr) {}
+
+ByamlIter::ByamlIter(const u8* data) {
+    mData = data;
+    mRootNode = nullptr;
+    if (mData) {
+        if (ByamlLocalUtil::verifiByaml(mData)) {
+            const ByamlHeader* header = reinterpret_cast<const ByamlHeader*>(mData);
+            u32 data_root_offset = header->getDataOffset();
+            if (data_root_offset) {
+                mRootNode = &mData[data_root_offset];
+            }
+        } else {
+            mData = nullptr;
+            mRootNode = nullptr;
+        }
+    }
+}
+
+ByamlIter::ByamlIter(const u8* data, const u8* root_node) : mData(data), mRootNode(root_node) {}
+
+ByamlIter::ByamlIter(const ByamlIter& other) : mData(other.mData), mRootNode(other.mRootNode) {}
+
+ByamlIter& ByamlIter::operator=(const ByamlIter& rhs) {
+    if (this != &rhs) {
+        mRootNode = rhs.mRootNode;
+        mData = rhs.mData;
+    }
+    return *this;
+}
+
+bool ByamlIter::isValid() const volatile {
+    return mData != nullptr;
+}
+
+bool ByamlIter::isTypeHash() const {
+    const ByamlContainerHeader* node = reinterpret_cast<const ByamlContainerHeader*>(mRootNode);
+    return node ? node->getType() == ByamlType::Hash : false;
+}
+
+bool ByamlIter::isTypeArray() const {
+    const ByamlContainerHeader* node = reinterpret_cast<const ByamlContainerHeader*>(mRootNode);
+    return node ? node->getType() == ByamlType::Array : false;
+}
+
+bool ByamlIter::isTypeContainer() const {
+    return isTypeArray() || isTypeHash();
+}
+
+bool ByamlIter::hasKey(const char* key) const {
+    if (!mRootNode) {
+        return false;
+    }
+
+    if (!isTypeHash()) {
+        return false;
+    }
+
+    const s32 index = getKeyIndex(key);
+    if (index < 0) {
+        return 0;
+    }
+
+    ByamlHashIter node_hash(mRootNode);
+    return node_hash.findPair(index);
+}
+
+s32 ByamlIter::getKeyIndex(const char* key) const {
+    const ByamlStringTableIter& table = ByamlLocalUtil::getHashKeyTableIter(mData);
+    return table.findStringIndex(key);
+}
+
+s32 ByamlIter::getSize() const {
+    if (!mRootNode) {
+        return 0;
+    }
+
+    if (isTypeContainer()) {
+        return ByamlLocalUtil::getContainerSize(mRootNode);
+    }
+
+    return 0;
+}
+
+ByamlIter ByamlIter::getIterByIndex(s32 index) const {
+    ByamlData data;
+    if (!getByamlDataByIndex(&data, index)) {
+        return ByamlIter();
+    }
+
+    return getIterFromData(data);
+}
+
+ByamlIter ByamlIter::getIterByKey(const char* key) const {
+    ByamlData data;
+    if (!getByamlDataByKey(&data, key)) {
+        return ByamlIter();
+    }
+
+    return getIterFromData(data);
+}
+
+ByamlIter ByamlIter::getIterFromData(const ByamlData& data) const {
+    const ByamlType& data_type = data.getType();
```

Comment:
```suggestion
    const ByamlType data_type = data.getType();
```

Reply:
Fails to match if I don't do this. A lot if the branches get's re-ordered.
It also matches if I directly use data.getType() directly in each location, which I guess it optimizes down into a const& value.

Reply:
Oh, I see. Could you use data.getType() directly? That would look a lot less weird than initializing a const-ref with a value and relying on [lifetime extension of temporaries](https://abseil.io/tips/107).

Reply:
Well, either way we're relying on this specific compiler optimizing it in that specific way, as in the assembly it only does the getType call once anyway. But sure.